Source/hu_stuff.c

```diff
@@ -435,6 +448,35 @@ void HU_Init(void)
         crosshair_nam[i] = NULL;
     }
   }
+
+  if (message_colorized)
+  {
+    // [crispy] colorize keycard and skull key messages
+    s_GOTBLUECARD = CrispyReplaceColor(s_GOTBLUECARD, CR_BLUE, " blue ");
+    s_GOTBLUESKUL = CrispyReplaceColor(s_GOTBLUESKUL, CR_BLUE, " blue ");
+    s_GOTREDCARD  = CrispyReplaceColor(s_GOTREDCARD,  CR_RED,  " red ");
+    s_GOTREDSKULL = CrispyReplaceColor(s_GOTREDSKULL, CR_RED,  " red ");
+    s_GOTYELWCARD = CrispyReplaceColor(s_GOTYELWCARD, CR_GOLD, " yellow ");
+    s_GOTYELWSKUL = CrispyReplaceColor(s_GOTYELWSKUL, CR_GOLD, " yellow ");
+    s_PD_BLUEC    = CrispyReplaceColor(s_PD_BLUEC,    CR_BLUE, " blue ");
+    s_PD_BLUEK    = CrispyReplaceColor(s_PD_BLUEK,    CR_BLUE, " blue ");
+    s_PD_BLUEO    = CrispyReplaceColor(s_PD_BLUEO,    CR_BLUE, " blue ");
+    s_PD_BLUES    = CrispyReplaceColor(s_PD_BLUES,    CR_BLUE, " blue ");
+    s_PD_REDC     = CrispyReplaceColor(s_PD_REDC,     CR_RED,  " red ");
+    s_PD_REDK     = CrispyReplaceColor(s_PD_REDK,     CR_RED,  " red ");
+    s_PD_REDO     = CrispyReplaceColor(s_PD_REDO,     CR_RED,  " red ");
+    s_PD_REDS     = CrispyReplaceColor(s_PD_REDS,     CR_RED,  " red ");
+    s_PD_YELLOWC  = CrispyReplaceColor(s_PD_YELLOWC,  CR_GOLD, " yellow ");
+    s_PD_YELLOWK  = CrispyReplaceColor(s_PD_YELLOWK,  CR_GOLD, " yellow ");
+    s_PD_YELLOWO  = CrispyReplaceColor(s_PD_YELLOWO,  CR_GOLD, " yellow ");
+    s_PD_YELLOWS  = CrispyReplaceColor(s_PD_YELLOWS,  CR_GOLD, " yellow ");
+
+    // [crispy] colorize multi-player messages
+    s_HUSTR_PLRGREEN  = CrispyReplaceColor(s_HUSTR_PLRGREEN,  CR_GREEN, "Green: ");
+    s_HUSTR_PLRINDIGO = CrispyReplaceColor(s_HUSTR_PLRINDIGO, CR_GRAY,  "Indigo: ");
+    s_HUSTR_PLRBROWN  = CrispyReplaceColor(s_HUSTR_PLRBROWN,  CR_BROWN, "Brown: ");
+    s_HUSTR_PLRRED    = CrispyReplaceColor(s_HUSTR_PLRRED,    CR_RED,   "Red: ");
+  }
```

Comment:
Can this be done in real time, without restarting the program?

Reply:
Right, another variant:
```diff
diff --git a/Source/hu_stuff.c b/Source/hu_stuff.c
index 6f34edf..24beb42 100644
--- a/Source/hu_stuff.c
+++ b/Source/hu_stuff.c
@@ -312,17 +312,90 @@ static boolean VANILLAMAP(int e, int m)
     return (e > 0 && e <= 4 && m > 0 && m <= 9);
 }
 
-static char* CrispyReplaceColor (const char *str, const int cr, const char *col)
+static char* CrispyReplaceColor (const char *str, const char *col)
 {
     char *str_replace, col_replace[16];
 
     M_snprintf(col_replace, sizeof(col_replace),
-               "\x1b%c%s\x1b%c", '0'+cr, col, '0'-1);
+               "\x1b %s\x1b%c", col, '0'-1);
     str_replace = M_StringReplace(str, col, col_replace);
 
     return str_replace;
 }
 
+static void UpdateColor(char *str, const int cr)
+{
+    int i;
+    int len = strlen(str);
+
+    for (i = 0; i < len; ++i)
+    {
+        if (str[i] == '\x1b' && i + 1 < len)
+        {
+          str[i + 1] = '0'+cr;
+          break;
+        }
+    }
+}
+
+void HU_ResetMessageColors(void)
+{
+    if (message_colorized)
+    {
+        // [crispy] colorize keycard and skull key messages
+        UpdateColor(s_GOTBLUECARD, CR_BLUE);
+        UpdateColor(s_GOTBLUESKUL, CR_BLUE);
+        UpdateColor(s_GOTREDCARD,  CR_RED);
+        UpdateColor(s_GOTREDSKULL, CR_RED);
+        UpdateColor(s_GOTYELWCARD, CR_GOLD);
+        UpdateColor(s_GOTYELWSKUL, CR_GOLD);
+        UpdateColor(s_PD_BLUEC,    CR_BLUE);
+        UpdateColor(s_PD_BLUEK,    CR_BLUE);
+        UpdateColor(s_PD_BLUEO,    CR_BLUE);
+        UpdateColor(s_PD_BLUES,    CR_BLUE);
+        UpdateColor(s_PD_REDC,     CR_RED);
+        UpdateColor(s_PD_REDK,     CR_RED);
+        UpdateColor(s_PD_REDO,     CR_RED);
+        UpdateColor(s_PD_REDS,     CR_RED);
+        UpdateColor(s_PD_YELLOWC,  CR_GOLD);
+        UpdateColor(s_PD_YELLOWK,  CR_GOLD);
+        UpdateColor(s_PD_YELLOWO,  CR_GOLD);
+        UpdateColor(s_PD_YELLOWS,  CR_GOLD);
+
+        // [crispy] colorize multi-player messages
+        UpdateColor(s_HUSTR_PLRGREEN,  CR_GREEN);
+        UpdateColor(s_HUSTR_PLRINDIGO, CR_GRAY);
+        UpdateColor(s_HUSTR_PLRBROWN,  CR_BROWN);
+        UpdateColor(s_HUSTR_PLRRED,    CR_RED);
+    }
+    else
+    {
+        UpdateColor(s_GOTBLUECARD, hudcolor_mesg);
+        UpdateColor(s_GOTBLUESKUL, hudcolor_mesg);
+        UpdateColor(s_GOTREDCARD,  hudcolor_mesg);
+        UpdateColor(s_GOTREDSKULL, hudcolor_mesg);
+        UpdateColor(s_GOTYELWCARD, hudcolor_mesg);
+        UpdateColor(s_GOTYELWSKUL, hudcolor_mesg);
+        UpdateColor(s_PD_BLUEC,    hudcolor_mesg);
+        UpdateColor(s_PD_BLUEK,    hudcolor_mesg);
+        UpdateColor(s_PD_BLUEO,    hudcolor_mesg);
+        UpdateColor(s_PD_BLUES,    hudcolor_mesg);
+        UpdateColor(s_PD_REDC,     hudcolor_mesg);
+        UpdateColor(s_PD_REDK,     hudcolor_mesg);
+        UpdateColor(s_PD_REDO,     hudcolor_mesg);
+        UpdateColor(s_PD_REDS,     hudcolor_mesg);
+        UpdateColor(s_PD_YELLOWC,  hudcolor_mesg);
+        UpdateColor(s_PD_YELLOWK,  hudcolor_mesg);
+        UpdateColor(s_PD_YELLOWO,  hudcolor_mesg);
+        UpdateColor(s_PD_YELLOWS,  hudcolor_mesg);
+
+        UpdateColor(s_HUSTR_PLRGREEN,  hudcolor_chat);
+        UpdateColor(s_HUSTR_PLRINDIGO, hudcolor_chat);
+        UpdateColor(s_HUSTR_PLRBROWN,  hudcolor_chat);
+        UpdateColor(s_HUSTR_PLRRED,    hudcolor_chat);
+    }
+}
+
 //
 // HU_Init()
 //
@@ -452,31 +525,33 @@ void HU_Init(void)
   if (message_colorized)
   {
     // [crispy] colorize keycard and skull key messages
-    s_GOTBLUECARD = CrispyReplaceColor(s_GOTBLUECARD, CR_BLUE, " blue ");
-    s_GOTBLUESKUL = CrispyReplaceColor(s_GOTBLUESKUL, CR_BLUE, " blue ");
-    s_GOTREDCARD  = CrispyReplaceColor(s_GOTREDCARD,  CR_RED,  " red ");
-    s_GOTREDSKULL = CrispyReplaceColor(s_GOTREDSKULL, CR_RED,  " red ");
-    s_GOTYELWCARD = CrispyReplaceColor(s_GOTYELWCARD, CR_GOLD, " yellow ");
-    s_GOTYELWSKUL = CrispyReplaceColor(s_GOTYELWSKUL, CR_GOLD, " yellow ");
-    s_PD_BLUEC    = CrispyReplaceColor(s_PD_BLUEC,    CR_BLUE, " blue ");
-    s_PD_BLUEK    = CrispyReplaceColor(s_PD_BLUEK,    CR_BLUE, " blue ");
-    s_PD_BLUEO    = CrispyReplaceColor(s_PD_BLUEO,    CR_BLUE, " blue ");
-    s_PD_BLUES    = CrispyReplaceColor(s_PD_BLUES,    CR_BLUE, " blue ");
-    s_PD_REDC     = CrispyReplaceColor(s_PD_REDC,     CR_RED,  " red ");
-    s_PD_REDK     = CrispyReplaceColor(s_PD_REDK,     CR_RED,  " red ");
-    s_PD_REDO     = CrispyReplaceColor(s_PD_REDO,     CR_RED,  " red ");
-    s_PD_REDS     = CrispyReplaceColor(s_PD_REDS,     CR_RED,  " red ");
-    s_PD_YELLOWC  = CrispyReplaceColor(s_PD_YELLOWC,  CR_GOLD, " yellow ");
-    s_PD_YELLOWK  = CrispyReplaceColor(s_PD_YELLOWK,  CR_GOLD, " yellow ");
-    s_PD_YELLOWO  = CrispyReplaceColor(s_PD_YELLOWO,  CR_GOLD, " yellow ");
-    s_PD_YELLOWS  = CrispyReplaceColor(s_PD_YELLOWS,  CR_GOLD, " yellow ");
+    s_GOTBLUECARD = CrispyReplaceColor(s_GOTBLUECARD, " blue ");
+    s_GOTBLUESKUL = CrispyReplaceColor(s_GOTBLUESKUL, " blue ");
+    s_GOTREDCARD  = CrispyReplaceColor(s_GOTREDCARD,  " red ");
+    s_GOTREDSKULL = CrispyReplaceColor(s_GOTREDSKULL, " red ");
+    s_GOTYELWCARD = CrispyReplaceColor(s_GOTYELWCARD, " yellow ");
+    s_GOTYELWSKUL = CrispyReplaceColor(s_GOTYELWSKUL, " yellow ");
+    s_PD_BLUEC    = CrispyReplaceColor(s_PD_BLUEC,    " blue ");
+    s_PD_BLUEK    = CrispyReplaceColor(s_PD_BLUEK,    " blue ");
+    s_PD_BLUEO    = CrispyReplaceColor(s_PD_BLUEO,    " blue ");
+    s_PD_BLUES    = CrispyReplaceColor(s_PD_BLUES,    " blue ");
+    s_PD_REDC     = CrispyReplaceColor(s_PD_REDC,     " red ");
+    s_PD_REDK     = CrispyReplaceColor(s_PD_REDK,     " red ");
+    s_PD_REDO     = CrispyReplaceColor(s_PD_REDO,     " red ");
+    s_PD_REDS     = CrispyReplaceColor(s_PD_REDS,     " red ");
+    s_PD_YELLOWC  = CrispyReplaceColor(s_PD_YELLOWC,  " yellow ");
+    s_PD_YELLOWK  = CrispyReplaceColor(s_PD_YELLOWK,  " yellow ");
+    s_PD_YELLOWO  = CrispyReplaceColor(s_PD_YELLOWO,  " yellow ");
+    s_PD_YELLOWS  = CrispyReplaceColor(s_PD_YELLOWS,  " yellow ");
 
     // [crispy] colorize multi-player messages
-    s_HUSTR_PLRGREEN  = CrispyReplaceColor(s_HUSTR_PLRGREEN,  CR_GREEN, "Green: ");
-    s_HUSTR_PLRINDIGO = CrispyReplaceColor(s_HUSTR_PLRINDIGO, CR_GRAY,  "Indigo: ");
-    s_HUSTR_PLRBROWN  = CrispyReplaceColor(s_HUSTR_PLRBROWN,  CR_BROWN, "Brown: ");
-    s_HUSTR_PLRRED    = CrispyReplaceColor(s_HUSTR_PLRRED,    CR_RED,   "Red: ");
+    s_HUSTR_PLRGREEN  = CrispyReplaceColor(s_HUSTR_PLRGREEN,  "Green: ");
+    s_HUSTR_PLRINDIGO = CrispyReplaceColor(s_HUSTR_PLRINDIGO, "Indigo: ");
+    s_HUSTR_PLRBROWN  = CrispyReplaceColor(s_HUSTR_PLRBROWN,  "Brown: ");
+    s_HUSTR_PLRRED    = CrispyReplaceColor(s_HUSTR_PLRRED,    "Red: ");
   }
+
+  HU_ResetMessageColors();
 }
 
 //
diff --git a/Source/hu_stuff.h b/Source/hu_stuff.h
index d7323b2..b89cfa5 100644
--- a/Source/hu_stuff.h
+++ b/Source/hu_stuff.h
@@ -62,6 +62,8 @@ void HU_Erase(void);
 
 boolean HU_DemoProgressBar(boolean force);
 
+void HU_ResetMessageColors(void);
+
 // killough 5/2/98: moved from m_misc.c:
 
 //jff 2/16/98 hud supported automap colors added
diff --git a/Source/m_menu.c b/Source/m_menu.c
index 2aec7f0..c917130 100644
--- a/Source/m_menu.c
+++ b/Source/m_menu.c
@@ -4164,8 +4164,8 @@ setup_menu_t mess_settings1[] =  // Messages screen
   {"Center Messages", S_YESNO, m_null, M_X,
    M_Y + mess_centered*M_SPC, {"message_centered"}},
 
-  {"Colorize Player Messages", S_YESNO|S_PRGWARN, m_null, M_X,
-   M_Y + mess_colorized*M_SPC, {"message_colorized"}},
+  {"Colorize Player Messages", S_YESNO, m_null, M_X,
+   M_Y + mess_colorized*M_SPC, {"message_colorized"}, 0, HU_ResetMessageColors},
 
   {"Message Color During Play", S_CRITEM, m_null, M_X,
    M_Y + mess_color_play*M_SPC, {"hudcolor_mesg"}, 0, NULL, hudcolor_str},
```

Reply:
M_StringReplace() always allocates a new string, so this will leak memory.

Reply:
> M_StringReplace() always allocates a new string, so this will leak memory.

But here I am not replacing the string, but only editing the color in the existing string.

Reply:
Oh, you are right. How about this for `UpdateColor()` instead of the huge branch in `HU_ResetMessageColors()`?

```patch
static void UpdateColor(char *str, const int cr)
{
    int i;
    int len = strlen(str);

    if (!message_colorized)
        cr = -1;

    for (i = 0; i < len; ++i)
    {
        if (str[i] == '\x1b' && i + 1 < len)
        {
          str[i + 1] = '0'+cr;
          break;
        }
    }
}
```

Reply:
> How about this for `UpdateColor()` instead of the huge branch in `HU_ResetMessageColors()`?

Sure, if it works (can't test right now).